contracts/VolOracle.sol

```diff
@@ -1,41 +1,64 @@
 // SPDX-License-Identifier: UNLICENSED
 pragma solidity >=0.8.4;
 
-import { Observation } from "@uniswap/v3-core/contracts/libraries/Oracle.sol";
 import { IUniswapV3Pool } from "@uniswap/v3-core/contracts/interfaces/IUniswapV3Pool.sol";
 
 contract VolOracle {
     // constructor
     // initializes a uniswap
 
+    struct VolObservation {
+        // the block timestamp of the observation
+        uint32 blockTimestamp;
+        // the tick accumulator, i.e. tick * time elapsed since the pool was first initialized
+        int56 tickCumulative;
+        // the tick square accumulator, i.e. tick * tick * time elapsed since the oracle was first initialized
+        uint112 tickSquareCumulative;
+        // whether or not the observation is initialized
+        bool initialized;
+    }
+
+
     // observation size
     // stores 30 days of data (with some buffer)
     uint256 public constant OBSERVATION_SIZE = 345600;
     uint256 public constant UNIV3_MAX_CARDINALITY = 65535;
+
+    struct VolOracleState {
+        // @dev Stores Observation arrays for each pool
+        // TODO: can we use constant here
+        VolObservation[] observations;
+        // @dev Stores lastBlockTimestamp when the observation was initialized for the pool
+        uint256 lastBlockTimestamp;
+        uint256 lastObservationIndex;
+    }
+
     // @dev Stores Observation arrays for each pool
-    mapping(address => Observation[UNIV3_MAX_CARDINALITY]) public observations;
-    // @dev Stores lastBlockTimestamp when the observation was initialized for the pool
-    mapping(address => uint256) public lastBlockTimestamps;
-    mapping(address => uint256) public lastObservationIndex;
+    mapping(address => VolOracleState) public oracleStates;
+
 
     function initPool(address _pool) external {
-        require(observations[_pool].length == 0, "Pool already initialized");
+        require(oracleStates[_pool].observations.length == 0, "Pool already initialized");
         // only initialize pools which have max cardinality
+        (, , uint16 observationIndex, uint16 observationCardinality, , , ) = IUniswapV3Pool(_pool).slot0();
         require(
-            IUniswapV3Pool(_pool).slot0.observationCardinality == UNIV3_MAX_CARDINALITY,
+            observationCardinality == UNIV3_MAX_CARDINALITY,
             "Pool not at max cardinality"
         );
         // TODO: if pool is not at max cardinality then grow it
         // initializing the pool to max size
-        observations[_pool] = new Observation[](OBSERVATION_SIZE);
-        lastBlockTimestamps[_pool] = block.timestamp;
-        lastObservationIndex[_pool] = IUniswapV3Pool(_pool).slot0.observationIndex;
+        oracleStates[_pool] = VolOracleState(
+            // TODO: can we initialize storage array here?
+            new VolObservation[](OBSERVATION_SIZE),
```

Comment:
what if we do something like this: https://github.com/Uniswap/v3-core/blob/main/contracts/UniswapV3Pool.sol#L99
```
new VolObservation[OBSERVATION_SIZE],
```